### PR TITLE
ENH / CoW: Add lazy copy to eval

### PR DESCRIPTION
**Pull request by phofl**
- [x] xref #49473 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/computation/eval.py [hunk 1]
```diff
@@ -373,7 +375,11 @@ def eval(
             # if returning a copy, copy only on the first assignment
             if not inplace and first_expr:
                 try:
-                    target = env.target.copy()
+                    target = env.target
+                    if isinstance(target, NDFrame):
+                        target = target.copy(deep=not using_copy_on_write())
```
Comment (jorisvandenbossche) [inline pandas/core/computation/eval.py:380]: You could also do `target.copy(deep=None)` here, which will have the same effect (without having to import `using_copy_on_write` here). I think this is how we did it in most places that had a `self.copy()` in the method

**Review by jorisvandenbossche — APPROVED** [review]

File: pandas/core/computation/eval.py (hunk 1 above)
Comment (phofl) [reply pandas/core/computation/eval.py]: Forgot about that, thx

**Comment by jorisvandenbossche** [comment]
The typing error is related? (don't see it on main, and it happens in the eval.py file, although I don't see how it actually can be related ..)

**Comment by jorisvandenbossche** [comment]
The ignore didn't seem to have worked ..

**Pull request merged**